_posts/2019-02-19-refactoring-state-pattern.markdown

```diff
@@ -0,0 +1,114 @@
+  def change_state
+    case state
+    when "cart"
+      state = "paying"
+    when "paying"
+      state = "payment_pending"
+    when "payment_pending"
+      state = "complete"
+    when "complete"
+      state = "delivered"
+    else "complete"
+      state = "cart"
+    end
+  end
+end
+```
+
+Doesn't feel good, right? What if you need to do some validations before allow to change to the next state? What if I need to go back to a previous state? The code would be more complex and the `change_state` method would clearly have more than one responsibility. We can solve this problem by applying the state pattern.
+
+There are a few gems that can help us with this job, but I recommend the [AASM gem](https://github.com/aasm/aasm) because it's easy to understand and implement, we just need to include the gem into our model and start to set our states:
+
+Back to Ombushop's code, using the assm we can refactor the `Order` model like this:
+
+```ruby
+class Order < ApplicationRecord
+  include AASM
+
+  aasm(column: 'state') do
+    state :cart, initial: true
+    state :checkout, before_enter: :update_totals!
+    state :pending
+    state :completed, after_enter: :finalize!
+    state :delivered
+    state :canceled
+    state :resumed,
+    state :returned,
+    state :paying
+
+    event :next do
+      transitions from: :cart, to: :checkout
+      transitions from: :checkout, to: :paying
+    end
+
+    event :pending do
+      transitions from: [:checkout, :paying], to: :pending
+    end
+
+    event :payment_received do
+      transitions from: :pending, to: :paid
+    end
+
+    event :cancel do
+      transitions to: :canceled
+    end
+
+    # ... several other events here
+  end
+```
+
+First of all we need to tell aasm what is the column used to save the state. Then we need to add all possible states (remember that we need to have a finite number of states) and set which one is the initial state. Them in the `event` blocks we can start to define the flow between the states. Simple like that.
+
+As you probably noticed we can also define callback functions to be called if we need to do any action before or after the object transition to some state.
+
+AASS also give some useful methods such like:
+
+```ruby
+  order = Order.new
+  order.pending? # => true
+  order.pending
+  order.may_complete? # => true
+  order.complete
+  order.completed? # => true
+```
+
+This can help us to manage the states in our code more efficiently without to create a lot of extra methods. It also makes easier to add a new state if you ever need in the future.
+
+## Conclusion
+Handle the states of an object is a hard job by itself. In this article we saw how apply the state pattern can make our life better offering a solution to this. Next time that you have an object that behaves differently depending on its current state, the number of states is finite and the object change of state frequently remember of this important design pattern.
```

Comment:
"Next time that you have an object that behaves differently depending on its current state, the number of states is finite and the object changeS state frequently, remember this important design pattern."